ENH: Allow specification of terms to fit in polynomial fitting functions

jonathanunderwood opened the pull request:
The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
ENH: Allow specification of terms to fit in legfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood commented:
Some rationale for this enhancement:

Frequently it's desirable to fit a restricted expansion in Legendre polynomials. For example, in physics, it's often the case that the symmetry of the problem under study means a distribution is well described by an expansion in only even L Legendre polynomials. This change allows that kind of restriction, and any others, to be carried out.

This change is designed to be entirely backwards compatible, and so shouldn't break existing code.

Let me know if you need anything else done inn order to merge this. Thanks in advance.

jonathanunderwood committed:
TST: Add tests for legfit with deg specified as list

jgunderwood commented:
Alright, we now have complete doc strings and test coverage for the expanded API, so I think we're good to go for a merge. Let me know if anything else is needed, and if you want me to clean up the commit history.

charris commented on numpy/polynomial/legendre.py:
```diff
@@ -1556,6 +1576,18 @@ def legfit(x, y, deg, rcond=None, full=False, w=None):
 
+    # Expand c to include non-fitted coefficients, which are set to
+    # zero
+    if restricted_L:
```
The insert function is not terribly efficient. It would be better to start with an array of zeros of appropriate shape and fill in the fitted coefficients.

I'm also not a fan of capitalization in variables if it can be avoided, which it usually can.

jonathanunderwood replied:
Interestingly, the original version (commit 21fb38f) did exactly as you suggested i.e. starting with an array of zeros. Easy enough to drop the patch that moves to the insert approach.

charris commented:
What about the 2d, 3d, and upcoming nd cases? See #6071. The specification of degrees in those cases will be a bit trickier.

From the "no one expects drudgery" department, we try to keep all 6 polynomial types synchronized. What goes for one, goes for all, which makes all these little enhancements turn into bigger projects.

At some point I'd like to modify lsq allow to linear restraints, which would provide another way to achieve this, but that is still in the nebulous future.

charris commented:
If this goes in, it will also need a mention in `doc/release/1.11.0-notes.rst`.

charris commented on numpy/polynomial/legendre.py:
```diff
@@ -1418,8 +1418,11 @@ def legfit(x, y, deg, rcond=None, full=False, w=None):
         passing in a 2D-array that contains one dataset per column.
-    deg : int
-        Degree of the fitting polynomial
+    deg : int or array_like
```
Should probably separate the new bits out so that the new bits can be annotated `.. versionadded:: 1.11.0`. Not sure what the best way to do that is.

jonathanunderwood replied:
How about adding a ..versionchanged:: 1.11.0 directive after the existing versionadded?

charris replied:
The `versionchanged` you refer to is the version in which the weights were added, so that doesn't work.

jonathanunderwood replied:
Ah, right, I had misunderstood the use of the directives, will fix it.

jonathanunderwood commented:
In reply to this:

> What about the 2d, 3d, and upcoming nd cases? See #6071. The specification of degrees in those cases will be a bit trickier.

I am not sure what you mean - there aren't fitting routines for the 2d, 3d, and nd cases - only evaluation functions. So, I am not sure what you're asking here?

charris commented:
Sound like a plan. Take a look at `doc/source/dev/gitwash/development_workflow.rst` for proper formatting of commit messages. The current commits should also be squashed. You can do all that with `git rebase -i HEAD~7`, where 7 is the current number of commits you want to work on, followed by a force push `git push --force origin HEAD`

jonathanunderwood commented:
OK, I've fixed up the versionchanged tag and squashed the commits down to 2 (one for legendre.py mods, and one for the tests). Can you let me know if you're happy with this for the legendre polynomials? If you are, I'll make equivalent commits for the other polynomials using the changes as a template.

jonathanunderwood committed:
ENH: Allow specification of terms to fit in chebfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for chebfit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in hermfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for hermfit with deg specified as list

charris commented:
Looks like it is headed in the right direction. There tends to be a lot of cut-n-paste in these things, so worth a careful check that the polynomial types are all correct. Note that when the feature is available for all polynomials, the documentation of the `fit` class method in `_polybase.py` needs an update with added tests for the polynomial classes in `test_classes.py`. It is also convention that the commit messages for the enhancements begin `ENH: `, and for the tests with `TST: `.

jonathanunderwood committed:
ENH: Allow specification of terms to fit in lagfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for lagfit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in polyfit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for polyfit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in fit method

jonathanunderwood committed:
TST: Add tests for check_fit with deg specified as list

jonathanunderwood committed:
ENH: Allow specification of terms to fit in hermefit

The argument `deg` is enhanced to allow an
array_like argument to past which specifies
which terms to include in the fit.

The returned coef array is exapnded to have
entries of 0 for all coefficients which were
not included in the fit.

jonathanunderwood committed:
TST: Add tests for hermefit with deg specified as list

jonathanunderwood commented:
Alright, as far as I can see this is all ready to merge - all tests pass without warnings or errors.

jonathanunderwood commented:
Rebased against master and added release note.

charris commented:
@jonathanunderwood  I'll get back to this as soon as 1.10.2rc2 is out.

homu commented:
:umbrella: The latest upstream changes (presumably #6851) made this pull request unmergeable. Please resolve the merge conflicts.

jonathanunderwood commented:
I'll resolve any merge conflicts when charris is ready to merge - the conflict will be the release note addition, which will keep becoming a merge conflict anytime someone updates the release notes.

charris commented:
I expect the conflicts are all in the release notes.

charris commented:
BTW, homu is a bot ;)

jonathanunderwood commented:
Hah, OK :).

charris commented:
You can rebase now, but with things going in for 1.11 you might need to do it again ;) Make a comment when you are ready as github doesn't send update notifications.

jonathanunderwood committed:
Add release note for polynomial fit deg changes

jonathanunderwood commented:
OK, rebased and ready to merge (I hope).

charris commented:
Thanks @jonathanunderwood . I (or you ;)may try to streamline the code and documentation a bit later, but let's get this in first.

jonathanunderwood commented:
Great thanks.

Yes, it did occur to me that there's massive opportunity for refactoring here. A generic fit function which accepts a vander functiuon as an argument would be where I'd start. That would remove a lot of duplication. Is that sort of refactoring something you're interested in?

charris commented:
Anything that streamlines the code is a plus. I'd probably change two other things. First check for 1-D arrays of deg rather than flatten the input. That is simpler and leads to more predictable results. You can use `array(deg, ndmin=1, copy=False)` for that. Sorting also provides the min and max values at `deg[0]` and `deg[-1]` and makes the result more deterministic. Second, in the documentation of `deg`, mention in the text that the list version is new in 1.11 rather than use the `.. versionchanged::` directive.

charris commented:
Thinking about it, I'd use `asarray(deg)` then check dimensions for scalar or 1-D and `dtype.kind` for `i`.

charris commented:
See #7052 for an update.
